### PR TITLE
finalizing some logic

### PR DESCRIPTION
**Pull request by uncomfyhalomacro**
- [x] Basic logging
- [x] Support for SrcTar
- [x] Support for SrcDir
- [x] Vendor and compressed vendored dependencies to tarball
- [ ] Support for multiple crate and manifest files

The last one is a bit hard to grasp but I think I have an idea now. Basically, we can use the `sync` option for crates that are not related or are not members of workspaces like from https://github.com/ibm-s390-linux/s390-tools. They may share one vendored tarball since they are `synced`. But this behavior of vendoring, I need to confirm if that works.

As for other features, I tried this with https://build.opensuse.org/package/show/Base:System/just

image

CC: @Firstyear

### CODE REVIEW

**Comment by uncomfyhalomacro** [comment]
I will squash this one as one commit once everything is finalized.

**Comment by uncomfyhalomacro** [comment]
I will add some conditional statements to check if files/directories exists.

File: src/bin/cargo_vendor.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        .tempdir()
+        .expect("Failed to create temporary working directory.");
+    let mut workdir: PathBuf = tmpdir.path().into();
+    debug!("Created temporary working directory: {:?}", workdir);
+
+    // One is required over the other and there can't be both anyway.
+    // NOTE: Because our struct `Opt` requires srctar or srcdir but not both, we put
+    // some `unreachable!` macros because they cannot be reached after all.
+    info!("Checking sources before vendor 🥡");
+    if args.srcdir.is_some() {
```
Comment (Firstyear) [inline src/bin/cargo_vendor.rs:60]: Could you do "if let Some(src) = args.srcdir" instead? Saves the subsequent let binding.

File: src/utils/compress.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    tar.append_dir_all(topdir, &src)?;
+    tar.finish()?;
+    info!(
+        "Successfully created Zstd compressed archive for {}",
+        src.to_string_lossy()
+    );
+    Ok(())
+}
+
+pub fn tarxz(
```
Comment (Firstyear) [inline src/utils/compress.rs:50]: doesn't cargo vendor also update the Cargo.lock? I know we make the tar's with "vendor" and "Cargo.lock" both in the top level dir, so I think this needs to take an iterator of path or similar so you can add multiple inputs to the tar.

File: src/utils/mod.rs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    Ok(srcdir.as_ref().into())
+}
+
+pub fn vendor(
+    opts: impl AsRef<Opts>,
+    prjdir: impl AsRef<Path>,
+    vendorname: Option<&str>,
+) -> Result<(), io::Error> {
+    let mut prjdir = prjdir.as_ref().to_path_buf();
+    info!(?prjdir);
```
Comment (Firstyear) [inline src/utils/mod.rs:66]: Could be debug or trace instead.

File: src/utils/mod.rs [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                String::from_utf8_unchecked(cargo_update.stderr)
+            });
+        } else {
+            info!("Successfully ran cargo update ❤️");
+        }
+    } else {
+        warn!("Disabled update of dependencies. You may reenable it for security updates.");
+    };
+
+    let cargo_vendor = process::Command::new("cargo")
```
Comment (Firstyear) [inline src/utils/mod.rs:104]: Could be a seperate function for clarity?

File: src/utils/mod.rs [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        });
+    } else {
+        info!(
+            "Generated cargo config from vendor with content:
+
+```
+{}
+```
+",
+            unsafe { std::str::from_utf8_unchecked(&cargo_vendor.stdout) }
```
Comment (Firstyear) [inline src/utils/mod.rs:126]: Try OsString instead of this?

File: src/utils/mod.rs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        );
+        fs::write(cargo_config, cargo_vendor.stdout)?;
+    };
+
+    info!("Proceeding to create compressed archive of vendored deps...");
+    prjdir.push("vendor/");
```
Comment (Firstyear) [inline src/utils/mod.rs:132]: Missing Cargo.lock, needed to be put in the tar if update=true.

**Review by Firstyear — COMMENTED** [review]
A few comments I think about what might be needed to get this production ready. Mostly we need to also define some templates of packages that we know are "edge" cases. I think wayshot is one, s390-tools is another, Kanidm is a chonky project and a good stress test, and maybe something else simple?

**Pull request merged**